In [1]:
%load_ext sql

In [2]:
import csv, sqlite3

con = sqlite3.connect('my_data1.db')
cur = con.cursor()

In [3]:
%sql sqlite:///my_data1.db

In [6]:
import pandas as pd
import numpy as pd

In [5]:
df = pd.read_csv('https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_2/data/Spacex.csv')


101

In [21]:
df.rename(columns = {'Landing _Outcome': 'landing_Outcome'}, inplace = True)

In [22]:
df.to_sql('SPACEXTBL', con, if_exists='replace', index=False, method = 'multi')

101

In [23]:
df.head()

Date Time (UTC) Booster_Version  Launch_Site  \
0  04-06-2010   18:45:00  F9 v1.0  B0003  CCAFS LC-40   
1  08-12-2010   15:43:00  F9 v1.0  B0004  CCAFS LC-40   
2  22-05-2012   07:44:00  F9 v1.0  B0005  CCAFS LC-40   
3  08-10-2012   00:35:00  F9 v1.0  B0006  CCAFS LC-40   
4  01-03-2013   15:10:00  F9 v1.0  B0007  CCAFS LC-40   

                                             Payload  PAYLOAD_MASS__KG_  \
0               Dragon Spacecraft Qualification Unit                  0   
1  Dragon demo flight C1, two CubeSats, barrel of...                  0   
2                              Dragon demo flight C2                525   
3                                       SpaceX CRS-1                500   
4                                       SpaceX CRS-2                677   

       Orbit         Customer Mission_Outcome      landing_Outcome  
0        LEO           SpaceX         Success  Failure (parachute)  
1  LEO (ISS)  NASA (COTS) NRO         Success  Failure (parachute)  
2  LEO (ISS)      NASA (COTS)         Success           No attempt  
3  LEO (ISS)       NASA (CRS)         Success           No attempt  
4  LEO (ISS)       NASA (CRS)         Success           No attempt

In [8]:
%sql select * from SPACEXTBL LIMIT 5;

 * sqlite:///my_data1.db
Done.


[('04-06-2010', '18:45:00', 'F9 v1.0  B0003', 'CCAFS LC-40', 'Dragon Spacecraft Qualification Unit', 0, 'LEO', 'SpaceX', 'Success', 'Failure (parachute)'),
 ('08-12-2010', '15:43:00', 'F9 v1.0  B0004', 'CCAFS LC-40', 'Dragon demo flight C1, two CubeSats, barrel of Brouere cheese', 0, 'LEO (ISS)', 'NASA (COTS) NRO', 'Success', 'Failure (parachute)'),
 ('22-05-2012', '07:44:00', 'F9 v1.0  B0005', 'CCAFS LC-40', 'Dragon demo flight C2', 525, 'LEO (ISS)', 'NASA (COTS)', 'Success', 'No attempt'),
 ('08-10-2012', '00:35:00', 'F9 v1.0  B0006', 'CCAFS LC-40', 'SpaceX CRS-1', 500, 'LEO (ISS)', 'NASA (CRS)', 'Success', 'No attempt'),
 ('01-03-2013', '15:10:00', 'F9 v1.0  B0007', 'CCAFS LC-40', 'SpaceX CRS-2', 677, 'LEO (ISS)', 'NASA (CRS)', 'Success', 'No attempt')]

In [24]:
%sql SELECT DISTINCT Landing_Outcome from SPACEXTBL;

 * sqlite:///my_data1.db
Done.


[('Failure (parachute)',),
 ('No attempt',),
 ('Uncontrolled (ocean)',),
 ('Controlled (ocean)',),
 ('Failure (drone ship)',),
 ('Precluded (drone ship)',),
 ('Success (ground pad)',),
 ('Success (drone ship)',),
 ('Success',),
 ('Failure',),
 ('No attempt ',)]

In [28]:
%sql SELECT * from SPACEXTBL LIMIT 5;

 * sqlite:///my_data1.db
Done.


[('04-06-2010', '18:45:00', 'F9 v1.0  B0003', 'CCAFS LC-40', 'Dragon Spacecraft Qualification Unit', 0, 'LEO', 'SpaceX', 'Success', 'Failure (parachute)'),
 ('08-12-2010', '15:43:00', 'F9 v1.0  B0004', 'CCAFS LC-40', 'Dragon demo flight C1, two CubeSats, barrel of Brouere cheese', 0, 'LEO (ISS)', 'NASA (COTS) NRO', 'Success', 'Failure (parachute)'),
 ('22-05-2012', '07:44:00', 'F9 v1.0  B0005', 'CCAFS LC-40', 'Dragon demo flight C2', 525, 'LEO (ISS)', 'NASA (COTS)', 'Success', 'No attempt'),
 ('08-10-2012', '00:35:00', 'F9 v1.0  B0006', 'CCAFS LC-40', 'SpaceX CRS-1', 500, 'LEO (ISS)', 'NASA (CRS)', 'Success', 'No attempt'),
 ('01-03-2013', '15:10:00', 'F9 v1.0  B0007', 'CCAFS LC-40', 'SpaceX CRS-2', 677, 'LEO (ISS)', 'NASA (CRS)', 'Success', 'No attempt')]

In [25]:
%sql SELECT Launch_Site FROM SPACEXTBL WHERE Launch_Site LIKE 'CCA%' LIMIT 5;

 * sqlite:///my_data1.db
Done.


[('CCAFS LC-40',),
 ('CCAFS LC-40',),
 ('CCAFS LC-40',),
 ('CCAFS LC-40',),
 ('CCAFS LC-40',)]

In [27]:
%sql SELECT SUM(PAYLOAD_MASS__KG_) AS total_payload FROM SPACEXTBL WHERE Customer LIKE 'NASA (CRS)';

 * sqlite:///my_data1.db
Done.


[(45596,)]

In [29]:
%sql SELECT avg(PAYLOAD_MASS__KG_) AS average_payload FROM SPACEXTBL WHERE Booster_Version LIKE 'F9 v1.1';

 * sqlite:///my_data1.db
Done.


[(2928.4,)]

In [30]:
%sql SELECT min(date) AS Early_Date FROM SPACEXTBL WHERE Landing_Outcome  LIKE 'Success (ground pad)';

 * sqlite:///my_data1.db
Done.


[('01-05-2017',)]

In [34]:
%sql SELECT DISTINCT Customer, Landing_Outcome,PAYLOAD_MASS__KG_ FROM SPACEXTBL WHERE Landing_Outcome ='Success (drone ship)' AND PAYLOAD_MASS__KG_ BETWEEN 4000 AND 6000;

 * sqlite:///my_data1.db
Done.


[('SKY Perfect JSAT Group', 'Success (drone ship)', 4696),
 ('SKY Perfect JSAT Group', 'Success (drone ship)', 4600),
 ('SES', 'Success (drone ship)', 5300),
 ('SES EchoStar', 'Success (drone ship)', 5200)]

In [38]:
%sql SELECT DISTINCT Mission_Outcome, Count(*) AS Numbers FROM SPACEXTBL GROUP BY Mission_Outcome;

 * sqlite:///my_data1.db
Done.


[('Failure (in flight)', 1),
 ('Success', 98),
 ('Success ', 1),
 ('Success (payload status unclear)', 1)]

In [41]:
%sql SELECT Booster_Version, Max_Payload FROM (SELECT Booster_Version, MAX(PAYLOAD_MASS__KG_) AS Max_Payload FROM SPACEXTBL GROUP BY Booster_Version) AS SUB;

 * sqlite:///my_data1.db
Done.


[('F9 B4  B1039.2', 2647),
 ('F9 B4  B1040.2', 5384),
 ('F9 B4  B1041.2', 9600),
 ('F9 B4  B1043.2', 6460),
 ('F9 B4 B1039.1', 3310),
 ('F9 B4 B1040.1', 4990),
 ('F9 B4 B1041.1', 9600),
 ('F9 B4 B1042.1', 3500),
 ('F9 B4 B1043.1', 5000),
 ('F9 B4 B1044', 6092),
 ('F9 B4 B1045.1', 362),
 ('F9 B4 B1045.2', 2697),
 ('F9 B5  B1046.1', 3600),
 ('F9 B5 B1046.2', 5800),
 ('F9 B5 B1046.3', 4000),
 ('F9 B5 B1046.4', 12050),
 ('F9 B5 B1047.2', 5300),
 ('F9 B5 B1047.3 ', 6500),
 ('F9 B5 B1048.2', 3000),
 ('F9 B5 B1048.3', 4850),
 ('F9 B5 B1048.4', 15600),
 ('F9 B5 B1048.5', 15600),
 ('F9 B5 B1049.2', 9600),
 ('F9 B5 B1049.3', 13620),
 ('F9 B5 B1049.4', 15600),
 ('F9 B5 B1049.5', 15600),
 ('F9 B5 B1049.6', 15440),
 ('F9 B5 B1049.7 ', 15600),
 ('F9 B5 B1051.2 ', 4200),
 ('F9 B5 B1051.3', 15600),
 ('F9 B5 B1051.4', 15600),
 ('F9 B5 B1051.5', 14932),
 ('F9 B5 B1051.6', 15600),
 ('F9 B5 B1056.2 ', 2268),
 ('F9 B5 B1056.3 ', 6956),
 ('F9 B5 B1056.4', 15600),
 ('F9 B5 B1058.2 ', 5500),
 ('F9 B5 B1058.3 ', 15600),
 ('F9 B5 B1058.4 ', 2972),
 ('F9 B5 B1059.2', 1977),
 ('F9 B5 B1059.3', 15410),
 ('F9 B5 B1059.4', 3130),
 ('F9 B5 B1060.2 ', 15600),
 ('F9 B5 B1060.3', 15600),
 ('F9 B5B1047.1', 7075),
 ('F9 B5B1048.1', 9600),
 ('F9 B5B1049.1', 7060),
 ('F9 B5B1050', 2500),
 ('F9 B5B1051.1', 12055),
 ('F9 B5B1054', 4400),
 ('F9 B5B1056.1 ', 2495),
 ('F9 B5B1058.1 ', 12530),
 ('F9 B5B1059.1', 2617),
 ('F9 B5B1060.1', 4311),
 ('F9 B5B1061.1 ', 12500),
 ('F9 B5B1062.1', 4311),
 ('F9 B5B1063.1', 1192),
 ('F9 FT  B1021.2', 5300),
 ('F9 FT  B1029.2', 3669),
 ('F9 FT  B1031.2', 5200),
 ('F9 FT  B1032.2', 4230),
 ('F9 FT  B1035.2', 2205),
 ('F9 FT  B1036.2', 9600),
 ('F9 FT  B1038.2', 2150),
 ('F9 FT B1019', 2034),
 ('F9 FT B1020', 5271),
 ('F9 FT B1021.1', 3136),
 ('F9 FT B1022', 4696),
 ('F9 FT B1023.1', 3100),
 ('F9 FT B1024', 3600),
 ('F9 FT B1025.1', 2257),
 ('F9 FT B1026', 4600),
 ('F9 FT B1029.1', 9600),
 ('F9 FT B1030', 5600),
 ('F9 FT B1031.1', 2490),
 ('F9 FT B1032.1', 5300),
 ('F9 FT B1034', 6070),
 ('F9 FT B1035.1', 2708),
 ('F9 FT B1036.1', 9600),
 ('F9 FT B1037', 6761),
 ('F9 FT B1038.1', 475),
 ('F9 v1.0  B0003', 0),
 ('F9 v1.0  B0004', 0),
 ('F9 v1.0  B0005', 525),
 ('F9 v1.0  B0006', 500),
 ('F9 v1.0  B0007', 677),
 ('F9 v1.1', 4535),
 ('F9 v1.1  B1003', 500),
 ('F9 v1.1 B1010', 2216),
 ('F9 v1.1 B1011', 4428),
 ('F9 v1.1 B1012', 2395),
 ('F9 v1.1 B1013', 570),
 ('F9 v1.1 B1014', 4159),
 ('F9 v1.1 B1015', 1898),
 ('F9 v1.1 B1016', 4707),
 ('F9 v1.1 B1017', 553),
 ('F9 v1.1 B1018', 1952)]

上下两个语句一样~~

In [43]:
%sql SELECT Booster_Version, MAX(PAYLOAD_MASS__KG_) AS Max_Payload FROM SPACEXTBL GROUP BY Booster_Version;

 * sqlite:///my_data1.db
Done.


[('F9 B4  B1039.2', 2647),
 ('F9 B4  B1040.2', 5384),
 ('F9 B4  B1041.2', 9600),
 ('F9 B4  B1043.2', 6460),
 ('F9 B4 B1039.1', 3310),
 ('F9 B4 B1040.1', 4990),
 ('F9 B4 B1041.1', 9600),
 ('F9 B4 B1042.1', 3500),
 ('F9 B4 B1043.1', 5000),
 ('F9 B4 B1044', 6092),
 ('F9 B4 B1045.1', 362),
 ('F9 B4 B1045.2', 2697),
 ('F9 B5  B1046.1', 3600),
 ('F9 B5 B1046.2', 5800),
 ('F9 B5 B1046.3', 4000),
 ('F9 B5 B1046.4', 12050),
 ('F9 B5 B1047.2', 5300),
 ('F9 B5 B1047.3 ', 6500),
 ('F9 B5 B1048.2', 3000),
 ('F9 B5 B1048.3', 4850),
 ('F9 B5 B1048.4', 15600),
 ('F9 B5 B1048.5', 15600),
 ('F9 B5 B1049.2', 9600),
 ('F9 B5 B1049.3', 13620),
 ('F9 B5 B1049.4', 15600),
 ('F9 B5 B1049.5', 15600),
 ('F9 B5 B1049.6', 15440),
 ('F9 B5 B1049.7 ', 15600),
 ('F9 B5 B1051.2 ', 4200),
 ('F9 B5 B1051.3', 15600),
 ('F9 B5 B1051.4', 15600),
 ('F9 B5 B1051.5', 14932),
 ('F9 B5 B1051.6', 15600),
 ('F9 B5 B1056.2 ', 2268),
 ('F9 B5 B1056.3 ', 6956),
 ('F9 B5 B1056.4', 15600),
 ('F9 B5 B1058.2 ', 5500),
 ('F9 B5 B1058.3 ', 15600),
 ('F9 B5 B1058.4 ', 2972),
 ('F9 B5 B1059.2', 1977),
 ('F9 B5 B1059.3', 15410),
 ('F9 B5 B1059.4', 3130),
 ('F9 B5 B1060.2 ', 15600),
 ('F9 B5 B1060.3', 15600),
 ('F9 B5B1047.1', 7075),
 ('F9 B5B1048.1', 9600),
 ('F9 B5B1049.1', 7060),
 ('F9 B5B1050', 2500),
 ('F9 B5B1051.1', 12055),
 ('F9 B5B1054', 4400),
 ('F9 B5B1056.1 ', 2495),
 ('F9 B5B1058.1 ', 12530),
 ('F9 B5B1059.1', 2617),
 ('F9 B5B1060.1', 4311),
 ('F9 B5B1061.1 ', 12500),
 ('F9 B5B1062.1', 4311),
 ('F9 B5B1063.1', 1192),
 ('F9 FT  B1021.2', 5300),
 ('F9 FT  B1029.2', 3669),
 ('F9 FT  B1031.2', 5200),
 ('F9 FT  B1032.2', 4230),
 ('F9 FT  B1035.2', 2205),
 ('F9 FT  B1036.2', 9600),
 ('F9 FT  B1038.2', 2150),
 ('F9 FT B1019', 2034),
 ('F9 FT B1020', 5271),
 ('F9 FT B1021.1', 3136),
 ('F9 FT B1022', 4696),
 ('F9 FT B1023.1', 3100),
 ('F9 FT B1024', 3600),
 ('F9 FT B1025.1', 2257),
 ('F9 FT B1026', 4600),
 ('F9 FT B1029.1', 9600),
 ('F9 FT B1030', 5600),
 ('F9 FT B1031.1', 2490),
 ('F9 FT B1032.1', 5300),
 ('F9 FT B1034', 6070),
 ('F9 FT B1035.1', 2708),
 ('F9 FT B1036.1', 9600),
 ('F9 FT B1037', 6761),
 ('F9 FT B1038.1', 475),
 ('F9 v1.0  B0003', 0),
 ('F9 v1.0  B0004', 0),
 ('F9 v1.0  B0005', 525),
 ('F9 v1.0  B0006', 500),
 ('F9 v1.0  B0007', 677),
 ('F9 v1.1', 4535),
 ('F9 v1.1  B1003', 500),
 ('F9 v1.1 B1010', 2216),
 ('F9 v1.1 B1011', 4428),
 ('F9 v1.1 B1012', 2395),
 ('F9 v1.1 B1013', 570),
 ('F9 v1.1 B1014', 4159),
 ('F9 v1.1 B1015', 1898),
 ('F9 v1.1 B1016', 4707),
 ('F9 v1.1 B1017', 553),
 ('F9 v1.1 B1018', 1952)]

In [44]:
%sql SELECT SUBSTR(Date,4,2) AS Month, Booster_Version, Launch_site FROM SPACEXTBL WHERE Landing_Outcome LIKE 'Failure%drone%' AND SUBSTR(Date,7,4) = '2015'

 * sqlite:///my_data1.db
Done.


[('01', 'F9 v1.1 B1012', 'CCAFS LC-40'),
 ('04', 'F9 v1.1 B1015', 'CCAFS LC-40')]

In [45]:
%sql SELECT Landing_Outcome, COUNT(*) AS Numbers FROM SPACEXTBL WHERE Landing_Outcome LIKE 'Success%' AND Date BETWEEN '04-06-2010' AND '20-03-2017' GROUP BY Landing_Outcome ORDER BY Numbers DESC;

 * sqlite:///my_data1.db
Done.


[('Success', 20), ('Success (drone ship)', 8), ('Success (ground pad)', 6)]

课后网站作业

In [56]:
%sql select distinct Launch_Site from spacextbl;

 * sqlite:///my_data1.db
Done.


[('CCAFS LC-40',), ('VAFB SLC-4E',), ('KSC LC-39A',), ('CCAFS SLC-40',)]

In [57]:
%sql select distinct Launch_Site as Launch_name, count(*) from spacextbl group by Launch_Site;

 * sqlite:///my_data1.db
Done.


[('CCAFS LC-40', 26),
 ('CCAFS SLC-40', 34),
 ('KSC LC-39A', 25),
 ('VAFB SLC-4E', 16)]


Overview of the DataSet
SpaceX has gained worldwide attention for a series of historic milestones.

It is the only private company ever to return a spacecraft from low-earth orbit, which it first accomplished in December 2010. SpaceX advertises Falcon 9 rocket launches on its website with a cost of 62 million dollars wheras other providers cost upward of 165 million dollars each, much of the savings is because Space X can reuse the first stage.

Therefore if we can determine if the first stage will land, we can determine the cost of a launch.

This information can be used if an alternate company wants to bid against SpaceX for a rocket launch.

This dataset includes a record for each payload carried during a SpaceX mission into outer space.

!pip3 install sqlalchemy
Requirement already satisfied: sqlalchemy in c:\users\qxj\anaconda3\lib\site-packages (1.3.9)
!pip3 install ipython-sql
Collecting ipython-sql
  Downloading ipython_sql-0.4.1-py3-none-any.whl (21 kB)
Requirement already satisfied: six in c:\users\qxj\anaconda3\lib\site-packages (from ipython-sql) (1.16.0)
Collecting prettytable<1
  Downloading prettytable-0.7.2.zip (28 kB)
Collecting sqlparse
  Downloading sqlparse-0.4.3-py3-none-any.whl (42 kB)
Requirement already satisfied: sqlalchemy>=0.6.7 in c:\users\qxj\anaconda3\lib\site-packages (from ipython-sql) (1.3.9)
Requirement already satisfied: ipython-genutils>=0.1.0 in c:\users\qxj\anaconda3\lib\site-packages (from ipython-sql) (0.2.0)
Requirement already satisfied: ipython>=1.0 in c:\users\qxj\anaconda3\lib\site-packages (from ipython-sql) (8.2.0)
Requirement already satisfied: setuptools>=18.5 in c:\users\qxj\anaconda3\lib\site-packages (from ipython>=1.0->ipython-sql) (61.2.0)
Requirement already satisfied: decorator in c:\users\qxj\anaconda3\lib\site-packages (from ipython>=1.0->ipython-sql) (5.1.1)
Requirement already satisfied: backcall in c:\users\qxj\anaconda3\lib\site-packages (from ipython>=1.0->ipython-sql) (0.2.0)
Requirement already satisfied: matplotlib-inline in c:\users\qxj\anaconda3\lib\site-packages (from ipython>=1.0->ipython-sql) (0.1.2)
Requirement already satisfied: prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0 in c:\users\qxj\anaconda3\lib\site-packages (from ipython>=1.0->ipython-sql) (3.0.20)
Requirement already satisfied: traitlets>=5 in c:\users\qxj\anaconda3\lib\site-packages (from ipython>=1.0->ipython-sql) (5.1.1)
Requirement already satisfied: jedi>=0.16 in c:\users\qxj\anaconda3\lib\site-packages (from ipython>=1.0->ipython-sql) (0.18.1)
Requirement already satisfied: colorama in c:\users\qxj\anaconda3\lib\site-packages (from ipython>=1.0->ipython-sql) (0.4.4)
Requirement already satisfied: pygments>=2.4.0 in c:\users\qxj\anaconda3\lib\site-packages (from ipython>=1.0->ipython-sql) (2.11.2)
Requirement already satisfied: stack-data in c:\users\qxj\anaconda3\lib\site-packages (from ipython>=1.0->ipython-sql) (0.2.0)
Requirement already satisfied: pickleshare in c:\users\qxj\anaconda3\lib\site-packages (from ipython>=1.0->ipython-sql) (0.7.5)
Requirement already satisfied: parso<0.9.0,>=0.8.0 in c:\users\qxj\anaconda3\lib\site-packages (from jedi>=0.16->ipython>=1.0->ipython-sql) (0.8.3)
Requirement already satisfied: wcwidth in c:\users\qxj\anaconda3\lib\site-packages (from prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0->ipython>=1.0->ipython-sql) (0.2.5)
Requirement already satisfied: asttokens in c:\users\qxj\anaconda3\lib\site-packages (from stack-data->ipython>=1.0->ipython-sql) (2.0.5)
Requirement already satisfied: pure-eval in c:\users\qxj\anaconda3\lib\site-packages (from stack-data->ipython>=1.0->ipython-sql) (0.2.2)
Requirement already satisfied: executing in c:\users\qxj\anaconda3\lib\site-packages (from stack-data->ipython>=1.0->ipython-sql) (0.8.3)
Building wheels for collected packages: prettytable
  Building wheel for prettytable (setup.py): started
  Building wheel for prettytable (setup.py): finished with status 'done'
  Created wheel for prettytable: filename=prettytable-0.7.2-py3-none-any.whl size=13714 sha256=e93a1fa1552c4fd816ae06d53642b1c1cd174fa41fba453bc2fe6dc450aea596
  Stored in directory: c:\users\qxj\appdata\local\pip\cache\wheels\75\f7\28\77a076f1fa8cbeda61aca712815d04d7a32435f04a26a2dd7b
Successfully built prettytable
Installing collected packages: sqlparse, prettytable, ipython-sql
Successfully installed ipython-sql-0.4.1 prettytable-0.7.2 sqlparse-0.4.3
Connect to the database
Let us first load the SQL extension and establish a connection with the database

%load_ext sql
The sql extension is already loaded. To reload it, use:
  %reload_ext sql
import csv, sqlite3
con = sqlite3.connect("my_data1.db")
cur = con.cursor()
import numpy as np
import pandas as pd
%sql sqlite:///my_data1.db # sql connect to database 
path = "/Users/QXJ/Desktop/IBM/SpaceX/Spacex.csv"
df = pd.read_csv(path, index_col = False)
df.head()
Date	Time (UTC)	Booster_Version	Launch_Site	Payload	PAYLOAD_MASS__KG_	Orbit	Customer	Mission_Outcome	Landing _Outcome
0	04-06-2010	18:45:00	F9 v1.0 B0003	CCAFS LC-40	Dragon Spacecraft Qualification Unit	0	LEO	SpaceX	Success	Failure (parachute)
1	08-12-2010	15:43:00	F9 v1.0 B0004	CCAFS LC-40	Dragon demo flight C1, two CubeSats, barrel of...	0	LEO (ISS)	NASA (COTS) NRO	Success	Failure (parachute)
2	22-05-2012	07:44:00	F9 v1.0 B0005	CCAFS LC-40	Dragon demo flight C2	525	LEO (ISS)	NASA (COTS)	Success	No attempt
3	08-10-2012	00:35:00	F9 v1.0 B0006	CCAFS LC-40	SpaceX CRS-1	500	LEO (ISS)	NASA (CRS)	Success	No attempt
4	01-03-2013	15:10:00	F9 v1.0 B0007	CCAFS LC-40	SpaceX CRS-2	677	LEO (ISS)	NASA (CRS)	Success	No attempt
# Reset index as columns 
df.reset_index(inplace=True)
df.rename(columns = {'Landing _Outcome':'Landing_Outcome'}, inplace = True)
df.columns
Index(['index', 'Date', 'Time (UTC)', 'Booster_Version', 'Launch_Site',
       'Payload', 'PAYLOAD_MASS__KG_', 'Orbit', 'Customer', 'Mission_Outcome',
       'Landing_Outcome'],
      dtype='object')
df.to_sql("SPACEXTBL", con, if_exists='replace', index=False, method="multi")
101
%sql SELECT * FROM SPACEXTBL LIMIT 5;
 * sqlite:///my_data1.db
Done.
index	Date	Time (UTC)	Booster_Version	Launch_Site	Payload	PAYLOAD_MASS__KG_	Orbit	Customer	Mission_Outcome	Landing_Outcome
0	04-06-2010	18:45:00	F9 v1.0 B0003	CCAFS LC-40	Dragon Spacecraft Qualification Unit	0	LEO	SpaceX	Success	Failure (parachute)
1	08-12-2010	15:43:00	F9 v1.0 B0004	CCAFS LC-40	Dragon demo flight C1, two CubeSats, barrel of Brouere cheese	0	LEO (ISS)	NASA (COTS) NRO	Success	Failure (parachute)
2	22-05-2012	07:44:00	F9 v1.0 B0005	CCAFS LC-40	Dragon demo flight C2	525	LEO (ISS)	NASA (COTS)	Success	No attempt
3	08-10-2012	00:35:00	F9 v1.0 B0006	CCAFS LC-40	SpaceX CRS-1	500	LEO (ISS)	NASA (CRS)	Success	No attempt
4	01-03-2013	15:10:00	F9 v1.0 B0007	CCAFS LC-40	SpaceX CRS-2	677	LEO (ISS)	NASA (CRS)	Success	No attempt
%sql SELECT DISTINCT Landing_Outcome from SPACEXTBL;
 * sqlite:///my_data1.db
Done.
Landing_Outcome
Failure (parachute)
No attempt
Uncontrolled (ocean)
Controlled (ocean)
Failure (drone ship)
Precluded (drone ship)
Success (ground pad)
Success (drone ship)
Success
Failure
No attempt
Write and execute SQL queries
1. Display the names of the unique launch sites in the space mission
%sql SELECT DISTINCT Launch_Site FROM SPACEXTBL;
 * sqlite:///my_data1.db
Done.
Launch_Site
CCAFS LC-40
VAFB SLC-4E
KSC LC-39A
CCAFS SLC-40
2.Display 5 records where launch sites begin with the string 'CCA'
%sql SELECT Launch_Site FROM SPACEXTBL WHERE Launch_Site LIKE 'CCA%' LIMIT 5;
 * sqlite:///my_data1.db
Done.
Launch_Site
CCAFS LC-40
CCAFS LC-40
CCAFS LC-40
CCAFS LC-40
CCAFS LC-40
3.Display the total payload mass carried by boosters launched by NASA (CRS)
%sql SELECT SUM(PAYLOAD_MASS__KG_) AS total_payload FROM SPACEXTBL WHERE Customer LIKE 'NASA (CRS)';
 * sqlite:///my_data1.db
Done.
total_payload
45596
4.Display average payload mass carried by booster version F9 v1.1
%sql SELECT avg(PAYLOAD_MASS__KG_) AS Avg_Payload FROM SPACEXTBL WHERE Booster_Version LIKE 'F9 v1.1';
 * sqlite:///my_data1.db
Done.
Avg_Payload
2928.4
5. List the date when the first succesful landing outcome in ground pad was acheived.
%sql SELECT min(date) AS Early_Date from SPACEXTBL where Landing_Outcome LIKE 'Success (ground pad)'
 * sqlite:///my_data1.db
Done.
Early_Date
01-05-2017
6. List the names of the boosters which have success in drone ship and have payload mass greater than 4000 but less than 6000
%sql SELECT DISTINCT Customer, Landing_Outcome,PAYLOAD_MASS__KG_ FROM SPACEXTBL WHERE Landing_Outcome ='Success (drone ship)' AND PAYLOAD_MASS__KG_ BETWEEN 4000 AND 6000;
 * sqlite:///my_data1.db
Done.
Customer	Landing_Outcome	PAYLOAD_MASS__KG_
SKY Perfect JSAT Group	Success (drone ship)	4696
SKY Perfect JSAT Group	Success (drone ship)	4600
SES	Success (drone ship)	5300
SES EchoStar	Success (drone ship)	5200
7. List the total number of successful and failure mission outcomes
%sql SELECT Mission_Outcome, Count(*) AS Numbers FROM SPACEXTBL GROUP BY Mission_Outcome;
 * sqlite:///my_data1.db
Done.
Mission_Outcome	Numbers
Failure (in flight)	1
Success	98
Success	1
Success (payload status unclear)	1
8.List the names of the booster_versions which have carried the maximum payload mass.
%sql SELECT Booster_Version, Max_Payload FROM (SELECT Booster_Version, MAX(PAYLOAD_MASS__KG_) AS Max_Payload FROM SPACEXTBL GROUP BY Booster_Version) AS Sub;
 * sqlite:///my_data1.db
Done.
Booster_Version	Max_Payload
F9 B4 B1039.2	2647
F9 B4 B1040.2	5384
F9 B4 B1041.2	9600
F9 B4 B1043.2	6460
F9 B4 B1039.1	3310
F9 B4 B1040.1	4990
F9 B4 B1041.1	9600
F9 B4 B1042.1	3500
F9 B4 B1043.1	5000
F9 B4 B1044	6092
F9 B4 B1045.1	362
F9 B4 B1045.2	2697
F9 B5 B1046.1	3600
F9 B5 B1046.2	5800
F9 B5 B1046.3	4000
F9 B5 B1046.4	12050
F9 B5 B1047.2	5300
F9 B5 B1047.3	6500
F9 B5 B1048.2	3000
F9 B5 B1048.3	4850
F9 B5 B1048.4	15600
F9 B5 B1048.5	15600
F9 B5 B1049.2	9600
F9 B5 B1049.3	13620
F9 B5 B1049.4	15600
F9 B5 B1049.5	15600
F9 B5 B1049.6	15440
F9 B5 B1049.7	15600
F9 B5 B1051.2	4200
F9 B5 B1051.3	15600
F9 B5 B1051.4	15600
F9 B5 B1051.5	14932
F9 B5 B1051.6	15600
F9 B5 B1056.2	2268
F9 B5 B1056.3	6956
F9 B5 B1056.4	15600
F9 B5 B1058.2	5500
F9 B5 B1058.3	15600
F9 B5 B1058.4	2972
F9 B5 B1059.2	1977
F9 B5 B1059.3	15410
F9 B5 B1059.4	3130
F9 B5 B1060.2	15600
F9 B5 B1060.3	15600
F9 B5B1047.1	7075
F9 B5B1048.1	9600
F9 B5B1049.1	7060
F9 B5B1050	2500
F9 B5B1051.1	12055
F9 B5B1054	4400
F9 B5B1056.1	2495
F9 B5B1058.1	12530
F9 B5B1059.1	2617
F9 B5B1060.1	4311
F9 B5B1061.1	12500
F9 B5B1062.1	4311
F9 B5B1063.1	1192
F9 FT B1021.2	5300
F9 FT B1029.2	3669
F9 FT B1031.2	5200
F9 FT B1032.2	4230
F9 FT B1035.2	2205
F9 FT B1036.2	9600
F9 FT B1038.2	2150
F9 FT B1019	2034
F9 FT B1020	5271
F9 FT B1021.1	3136
F9 FT B1022	4696
F9 FT B1023.1	3100
F9 FT B1024	3600
F9 FT B1025.1	2257
F9 FT B1026	4600
F9 FT B1029.1	9600
F9 FT B1030	5600
F9 FT B1031.1	2490
F9 FT B1032.1	5300
F9 FT B1034	6070
F9 FT B1035.1	2708
F9 FT B1036.1	9600
F9 FT B1037	6761
F9 FT B1038.1	475
F9 v1.0 B0003	0
F9 v1.0 B0004	0
F9 v1.0 B0005	525
F9 v1.0 B0006	500
F9 v1.0 B0007	677
F9 v1.1	4535
F9 v1.1 B1003	500
F9 v1.1 B1010	2216
F9 v1.1 B1011	4428
F9 v1.1 B1012	2395
F9 v1.1 B1013	570
F9 v1.1 B1014	4159
F9 v1.1 B1015	1898
F9 v1.1 B1016	4707
F9 v1.1 B1017	553
F9 v1.1 B1018	1952
9. List the records which will display the month names, failure landing_outcomes in drone ship ,booster versions, launch_site for the months in year 2015.
Note: SQLLite does not support monthnames. So you need to use substr(Date, 4, 2) as month to get the months and substr(Date,7,4)='2015' for year.

%sql SELECT SUBSTR(Date,4,2) AS Month, Booster_Version, Launch_site FROM SPACEXTBL WHERE Landing_Outcome LIKE 'Failure%drone%' AND SUBSTR(Date,7,4) = '2015'
 * sqlite:///my_data1.db
Done.
Month	Booster_Version	Launch_Site
01	F9 v1.1 B1012	CCAFS LC-40
04	F9 v1.1 B1015	CCAFS LC-40
10. Rank the count of successful landing_outcomes between the date 04-06-2010 and 20-03-2017 in descending order.
%sql SELECT Landing_Outcome, COUNT(*) AS Numbers FROM SPACEXTBL WHERE Landing_Outcome LIKE 'Success%' AND Date BETWEEN '04-06-2010' AND '20-03-2017' GROUP BY Landing_Outcome ORDER BY Numbers DESC;
 * sqlite:///my_data1.db
Done.
Landing_Outcome	Numbers
Success	20
Success (drone ship)	8
Success (ground pad)	6
 